src/OSSupport/UDPEndpointImpl.cpp

```diff
@@ -0,0 +1,608 @@
+		if (m_IsMainSockIPv6)
+		{
+			if (m_HasIPv6)
+			{
+				sendto(m_MainSock, m_Data.data(), static_cast<socklen_t>(m_Data.size()), 0, reinterpret_cast<const sockaddr *>(&m_AddrIPv6), static_cast<socklen_t>(sizeof(m_AddrIPv6)));
+			}
+			else if (m_HasIPv4)
+			{
+				// If the secondary socket is valid, it is an IPv4 socket, so use that:
+				if (m_SecondSock != -1)
+				{
+					sendto(m_SecondSock, m_Data.data(), static_cast<socklen_t>(m_Data.size()), 0, reinterpret_cast<const sockaddr *>(&m_AddrIPv4), static_cast<socklen_t>(sizeof(m_AddrIPv4)));
+				}
+				else
+				{
+					// Need an address conversion from IPv4 to IPv6-mapped-IPv4:
+					ConvertIPv4ToMappedIPv6(m_AddrIPv4, m_AddrIPv6);
+					sendto(m_MainSock, m_Data.data(), static_cast<socklen_t>(m_Data.size()), 0, reinterpret_cast<const sockaddr *>(&m_AddrIPv6), static_cast<socklen_t>(sizeof(m_AddrIPv6)));
+				}
+			}
+			else
+			{
+				LOGD("UDP endpoint queued sendto: Name not resolved");
+				return;
+			}
+		}
+		else  // m_IsMainSockIPv6
+		{
+			// Main socket is IPv4 only, only allow IPv4 dst address:
+			if (!m_HasIPv4)
+			{
+				LOGD("UDP endpoint queued sendto: Name not resolved to IPv4 for an IPv4-only socket");
+				return;
+			}
+			sendto(m_MainSock, m_Data.data(), static_cast<socklen_t>(m_Data.size()), 0, reinterpret_cast<const sockaddr *>(&m_AddrIPv4), static_cast<socklen_t>(sizeof(m_AddrIPv4)));
+		}
+	}
+
+	virtual void OnError(int a_ErrorCode, const AString & a_ErrorMsg) override
+	{
+		// Nothing needed
+	}
+};
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// cUDPEndpointImpl:
+
+cUDPEndpointImpl::cUDPEndpointImpl(UInt16 a_Port, cUDPEndpoint::cCallbacks & a_Callbacks):
+	super(a_Callbacks),
+	m_Port(0),
+	m_MainSock(-1),
+	m_IsMainSockIPv6(true),
+	m_SecondarySock(-1),
+	m_MainEvent(nullptr),
+	m_SecondaryEvent(nullptr)
+{
+	Open(a_Port);
+}
+
+
+
+
+
+void cUDPEndpointImpl::Close(void)
+{
+	if (m_Port == 0)
+	{
+		// Already closed
+		return;
+	}
+
+	// Close the LibEvent handles:
+	if (m_MainEvent != nullptr)
+	{
+		event_free(m_MainEvent);
+		m_MainEvent = nullptr;
+	}
+	if (m_SecondaryEvent != nullptr)
+	{
+		event_free(m_SecondaryEvent);
+		m_SecondaryEvent = nullptr;
+	}
+
+	// Close the OS sockets:
+	evutil_closesocket(m_MainSock);
+	m_MainSock = -1;
+	evutil_closesocket(m_SecondarySock);
+	m_SecondarySock = -1;
+
+	// Mark as closed:
+	m_Port = 0;
+}
+
+
+
+
+
+bool cUDPEndpointImpl::IsOpen(void) const
+{
+	return (m_Port != 0);
+}
+
+
+
+
+
+UInt16 cUDPEndpointImpl::GetPort(void) const
+{
+	return m_Port;
+}
+
+
+
+
+
+bool cUDPEndpointImpl::Send(const AString & a_Payload, const AString & a_Host, UInt16 a_Port)
+{
+	// If a_Host is an IP address, send the data directly:
+	sockaddr_storage sa;
+	socklen_t salen = static_cast<int>(sizeof(sa));
+	memset(&sa, 0, sizeof(sa));
+	if (evutil_parse_sockaddr_port(a_Host.c_str(), reinterpret_cast<sockaddr *>(&sa), &salen) != 0)
```

Comment:
No, it's there, it's just a type mismatch - int vs unsigned vs size_t. These APIs are a royal PITA to work with multi-platformly.
Weird thing is, the compilation had already succeeded in a previous build: https://travis-ci.org/mc-server/MCServer/builds/51497191